pfsimulator/amps/mpi1/amps_exchange.c

```diff
@@ -28,11 +28,126 @@
 
 #include "amps.h"
 
-#ifdef AMPS_MPI_NOT_USE_PERSISTENT
+/* This CUDA stuff could be combined with AMPS_MPI_NOT_USE_PERSISTENT case */
+#ifdef PARFLOW_HAVE_CUDA
+
+void _amps_wait_exchange(amps_Handle handle)
+{
+  char *combuf;
+  int errchk;
+  int i;
+  int size;
+  MPI_Status *status;
+
+  if (handle->package->num_recv + handle->package->num_send)
+  {
+    status = (MPI_Status*)calloc((handle->package->num_recv +
+                                  handle->package->num_send), sizeof(MPI_Status));
+
+    MPI_Waitall(handle->package->num_recv + handle->package->num_send,
+                handle->package->recv_requests, status);
+    for (i = 0; i < handle->package->num_recv; i++)
+    {
+      errchk = amps_gpupacking(AMPS_UNPACK, 
+                 handle->package->recv_invoices[i], 
+                   i, &combuf, &size);
+      // if(errchk) printf("GPU unpacking failed at line: %d\n", errchk);
+    }
+    for (i = 0; i < handle->package->num_recv; i++)
+    {
+      amps_gpu_sync_streams(i);
+      AMPS_CLEAR_INVOICE(handle->package->recv_invoices[i]);
+    }
+    free(status);
```

Comment:
This should be moved up to the if clause.   If status is not NULL initialized I think this is a potential bug.

Reply:
Could you clarify this a bit more, maybe give line numbers what should be moved and where?

Reply:
I am still confused about this, but I removed the recurring alloc/dealloc, and the alloc/dealloc is now done when calling amps_NewPackage/amps_FreePackage.

Reply:
OK I got messed up in scoping, thought the free(status) was outside the L42 if.